If the text of a control is not navigable, use NVDAObjectTextInfo for object review and read line.

jcsteh opened the pull request:
### Link to issue number:
Fixes #15159. Fixes #11285.

### Summary of the issue:
In focus mode in web browsers, when a control specifies a label using aria-label or aria-labelledby, it is impossible to review or spell that label; e.g. using read current line or the review cursor. This is particularly annoying and detrimental to efficiency when using web apps such as Gmail and Slack, among others, where you have to switch to browse mode to work around this even though it might be more efficient to use focus mode for navigation.

The same bug also applies in Google Chrome menus and dialogs where UIA is not used, such as in Chrome 137.

### Description of user facing changes:
In focus mode in web browsers, it is now possible to review and spell the labels of controls when those labels are specifically provided for accessibility; e.g. via aria-label or aria-labelledby.

### Description of development approach:
When in focus mode, for non-editable controls, both speech and braille report the name, description, value, etc. of the control; they don't use the TextInfo. However, previously, the review cursor and the read line command always used the TextInfo. Since IA2TextTextInfo is used for all objects supporting IAccessibleText, this meant that if the content wasn't a flat piece of text equal to the label, the review cursor and read line command would report something different, often nothing at all. To fix this, use NVDAObjectTextInfo instead of the object's TextInfo for object review and the read line command.

### Testing strategy:
In both Firefox and Edge:

1. Followed and verified the steps to reproduce in https://github.com/nvaccess/nvda/issues/15159#issue-1808844106.
2. Also tested with the review cursor in the Gmail message list, where previously, the review cursor couldn't read the message details. With this change, it can. Same for read current line.
3. Verified that braille remains the same as before (the expected behaviour) in all of these scenarios.
4. Also tested this test case:
    `data:text/html,<textarea>ab`
    Verified that in focus mode, formatting info is still reported for the review cursor, indicating that the correct TextInfo is being used. Also verified that the review cursor tracks the caret.
5. With a multi line message in Gmail, verified that read current line reports the correct line when on the second or subsequent lines.
6. I wasn't able to reproduce #11285 myself. However, @amirsol81 [verified that this change fixes that too](https://github.com/nvaccess/nvda/issues/11285#issuecomment-2996421115).

### Known issues with pull request:
Although this fix is primarily about focus mode in web browsers, the code paths it touches are more generic than that. Anything that has a specific TextInfo but for which `_hasNavigableText` returns False will be impacted. This is consistent with how we report the focus with speech and what is shown in braille, and thus I think this makes sense. Nevertheless, this potential broader impact is worth noting.

### Code Review Checklist:

- [x] Documentation:
  - Change log entry
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] UX of all users considered:
  - Speech
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] API is compatible with existing add-ons.
- [x] Security precautions taken.

<!-- Please keep the following -->
@coderabbitai summary

jcsteh commented:
Converting to draft until I figure out the Chrome system test failures.

cary-rowen commented:
Just a reminder: if this works as expected #17573, #16248 can be revert.

jcsteh commented:
Unfortunately, I have to go back to square 0 on this. It breaks reporting of annotation details. I have some thoughts on another solution using `_hasNavigableText`, but I think it's going to be a bit more complicated than this current patch.

jcsteh committed:
If the text of a control is not navigable, use NVDAObjectTextInfo for object review and read line.

seanbudd commented on user_docs/en/changes.md:
```diff
@@ -34,6 +34,7 @@
   * Search suggestions are now reported reliably.
 * In Windows 11, NVDA will no longer announce emoji panel items twice while browsing them. (#18236, @josephsl)
 * It is now possible to route to any braille cell on the Humanware Monarch multiline braille device, using their point and click action. (#18248)
+* In focus mode in web browsers, it is now possible to review and spell the labels of controls when those labels are specifically provided for accessibility; e.g. via aria-label or aria-labelledby. (#15159, @jcsteh)
```
```suggestion
* In focus mode in web browsers, it is now possible to review and spell the labels of controls when those labels are specifically provided for accessibility; e.g. via `aria-label` or `aria-labelledby`. (#15159, @jcsteh)
```

seanbudd commented on source/review.py:
```diff
@@ -28,18 +28,22 @@ def getObjectPosition(obj):
+				pos = obj.makeTextInfo(textInfos.POSITION_FIRST)
+			except (NotImplementedError, RuntimeError):
+				log.debugWarning(
+					"%s does not support POSITION_FIRST, falling back to NVDAObjectTextInfo" % obj.TextInfo,
```
```suggestion
					f"{obj.TextInfo} does not support POSITION_FIRST, falling back to NVDAObjectTextInfo",
```

seanbudd commented on source/review.py:
```diff
@@ -28,18 +28,22 @@ def getObjectPosition(obj):
-		pos = obj.makeTextInfo(textInfos.POSITION_CARET)
-	except (NotImplementedError, RuntimeError):
-		# No caret supported, try first position instead
+	isNavigable: bool = obj._hasNavigableText
```
Can you update this function to add type hints and remove them from the docstring?

jcsteh commented:
I can't reproduce #11285 even with NVDA alpha.

jcsteh committed:
Apply suggestions from code review

Co-authored-by: Sean Budd <seanbudd123@gmail.com>

jcsteh committed:
Annotations

jcsteh committed:
Update changes to note that this also fixes #11285.

jcsteh commented:
I'm not willing to take on the reverts of the Visual Studio Code changes, as I don't use Visual Studio Code here and thus can't test whether this more generic fix actually addresses those issues correctly. However, someone else is very welcome to pursue that if/when this PR gets merged.

seanbudd committed:
Apply suggestions from code review

seanbudd commented:
Thanks @jcsteh
